subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,146 +122,172 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
+			      const uint8_t *metadata, size_t metadata_len)
 {
-	uint16_t idx;
-	bool prev;
-	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+	if (metadata_len > CONFIG_BT_MESH_DFU_METADATA_MAXLEN) {
+		return -EFBIG;
 	}
 
-	idx = slot_idx(slot);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
+	}
 
-	LOG_DBG("%u: %u", idx, valid);
+	slot->slot.size = size;
+	slot->slot.metadata_len = metadata_len;
+	memcpy(slot->slot.metadata, metadata, metadata_len);
+	return 0;
+}
 
-	if (valid) {
-		prev = atomic_test_and_set_bit(valid_slots, idx);
-	} else {
-		prev = atomic_test_and_clear_bit(valid_slots, idx);
-	}
+int bt_mesh_dfu_slot_fwid_set(struct bt_mesh_dfu_slot *dfu_slot,
+			      const uint8_t *fwid, size_t fwid_len)
+{
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (valid == prev) {
-		return 0;
+	if (fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		return -EFBIG;
 	}
 
-	err = valid_slots_store();
-	if (err) {
-		LOG_WRN("Storage failed. err: %d", err);
-		atomic_set_bit_to(valid_slots, idx, prev);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	return err;
+	slot->slot.fwid_len = fwid_len;
+	memcpy(slot->slot.fwid, fwid, fwid_len);
+	return 0;
 }
 
-bool bt_mesh_dfu_slot_is_valid(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_commit(struct bt_mesh_dfu_slot *dfu_slot)
 {
-	uint16_t idx;
+	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_RESERVED ||
+	    slot->slot.fwid_len == 0 ||
+	    slot->slot.size == 0) {
+		return -EINVAL;
+	}
+
+	if (bt_mesh_dfu_slot_get(slot->slot.fwid, slot->slot.fwid_len, NULL)
+	    >= 0) {
+		return -EALREADY;
+	}
+
+	slot->idx = committed_slot_count;
+	slot->state = SLOT_STATE_COMMITTED;
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return false;
+	err = slot_store(slot);
+	if (err) {
+		LOG_WRN("Store failed (err: %d)", err);
+		slot->state = SLOT_STATE_RESERVED;
+		slot->idx = 0;
+		return err;
 	}
 
-	idx = slot_idx(slot);
-	return atomic_test_bit(valid_slots, idx);
+	committed_slot_count++;
+
+	LOG_DBG("Stored slot #%u: %s", slot - &slots[0],
+		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
+	return 0;
 }
 
-int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_release(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	struct slot *s = CONTAINER_OF(slot, struct slot, slot);
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	LOG_DBG("%u", s - &slots[0]);
-
-	slot_erase(s);
-	slot_invalidate(s);
-	sys_slist_find_and_remove(&list, &s->n);
+	slot->state = SLOT_STATE_FREE;
 
 	return 0;
 }
 
-int bt_mesh_dfu_slot_del_all(void)
+int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *dfu_slot)
```

Comment:
Can you check ideompotencay of important slot APIs in the test. Like "del", "release", "add", and "del_all". Calling these APIs multiple times with same params, should not result in misbehavior.